src/ast/desugar.rs

```diff
@@ -375,11 +383,77 @@ pub(crate) fn desugar_command(
             if name == "".into() {
                 name = rule.to_string().replace('\"', "'").into();
             }
-            vec![NCommand::NormRule {
-                ruleset,
-                name,
-                rule: flatten_rule(rule, desugar),
-            }]
+
+            if seminaive {
+                let mut new_rule = rule.clone();
+                // only add new rule when there is Call in body to avoid adding same rule.
+                let mut add_new_rule = false;
+
+                for head_slice in new_rule.head.iter_mut() {
+                    match head_slice {
+                        Action::Set(_, _, value) => {
+                            // if the right hand side is a function call,
+                            // move it to body so seminaive fires
+                            match value {
+                                Expr::Call(_, _) => {
+                                    add_new_rule = true;
+                                    let mut eq_vec: Vec<Expr> = Vec::new();
+                                    let fresh_symbol = desugar.get_fresh();
+                                    eq_vec.push(Expr::Var(fresh_symbol));
+                                    eq_vec.push(value.clone());
+                                    new_rule.body.push(Fact::Eq(eq_vec));
+                                    *value = Expr::Var(fresh_symbol);
+                                }
+                                _ => (),
+                            };
+                        }
+
+                        // move let binding to body.
+                        Action::Let(symbol, expr) => {
+                            let mut eq_vec: Vec<Expr> = Vec::new();
+                            eq_vec.push(Expr::Var(symbol.clone()));
+                            eq_vec.push(expr.clone());
+                            new_rule.body.push(Fact::Eq(eq_vec));
+                        }
+                        _ => (),
+                    }
+                }
+
+                // remove all let action
+                new_rule.head.retain_mut(|action| match action {
```

Comment:
`|action| matches!(action, Action::Let(_))` might be more concise. Also maybe considering put the retaining inside the `if add_new_rule { ... }` block?